### DISCUSSION TITLE
How to properly export Mongoose Models and use in Next.js?

### DISCUSSION THREAD

**Discussion by nicopoore** [vercel/next.js#21500] [Help]
Hey, I'm writing my first Next.js app with TypeScript and was running into an issue with the standard way of writing and exporting Mongoose Schemas and Models: exporting with `export default mongoose.model('[Model]', [model]Schema, '[models]')` was, for some reason unknown to me, making the model be overwritten every time it was used or called. 

After some searching, I found that on the "official" Next.js+MongoDB+Mongoose example, the solution to this was to instead export with `export default mongoose.models.[Model] || mongoose.model('[Model]', [model]Schema, '[models]')`, making it only create a model if there wasn't one already. I had seen this solution in a stackoverflow thread but it seemed more like a bandaid solution than a fix to the problem so I ignored it, but since it was on the official examples, I now assumed that was the correct way of importing Models in Next.js.

This worked fine until today, when I tried to do [Model].findByIdAndUpdate(). TypeScript instantly recognized an error, which after some time I found out was coming from the Model export using that weird statement instead of the usual `export default mongoose.model(...)`. I haven't found a way to fix this other than straight up making an exception in TypeScript which again seems like just putting a bandaid on a deeper problem.

Am I doing something wrong, or is there something I'm missing?

Thanks in advance!

**Comment by lafisrap** [comment]
Hi @nicopoore , I also tried the mongo team solution, but then found also this

```
let ZoomToken;
try {
    ZoomToken = mongoose.model<Interface>('ZoomToken');
} catch (error) {
    ZoomToken = mongoose.model<Interface>('ZoomToken', Model);
}
```

which seems to be working and gives a better feeling as to use a somewhat unofficial feature.

**Comment by landermedia** [comment]
I'm curious as well as to how to solve this. I just ran into this today and I guess I'm unable to find a solution of fixing the correct type. Invoking `Download.findOneAndUpdate` isn't now valid.

**Comment by dee2025** [comment]
export default mongoose.models.[ Model ] || mongoose.model ( '[Model]', Model-Schema );

**Comment by RadotJar** [comment]
Also curious to hear the official word on this.
